File: tests/org/elixir_lang/reference/callable/Issue692Test.java
package org.elixir_lang.reference.callable;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import com.intellij.testFramework.fixtures.LightCodeInsightFixtureTestCase;
import org.elixir_lang.psi.ElixirIdentifier;
import org.elixir_lang.psi.call.Call;

public class Issue692Test extends LightCodeInsightFixtureTestCase {
    /*
     * Tests
     */

    public void testUnresolvedAtTopOfFile() {
        myFixture.configureByFiles("unresolved_at_top_of_file.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);
        assertInstanceOf(elementAtCaret, LeafPsiElement.class);

        PsiElement parent = elementAtCaret.getParent();

        assertNotNull(parent);
        assertInstanceOf(parent, ElixirIdentifier.class);

        PsiElement grandParent = parent.getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        Call grandParentCall = (Call) grandParent;

        PsiReference reference = grandParentCall.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNotNull(resolved);
        assertTrue(resolved.isEquivalentTo(grandParent));
    }

    /*
     * Protected Instance Methods
     */

    @Override
    protected String getTestDataPath() {
        return "testData/org/elixir_lang/reference/callable/issue_692";
    }
}

File: tests/org/elixir_lang/reference/callable/Issue692Test.kt
package org.elixir_lang.reference.callable

import org.elixir_lang.psi.ElixirIdentifier
import com.intellij.psi.impl.source.tree.LeafPsiElement
import org.elixir_lang.PlatformTestCase
import org.elixir_lang.psi.call.Call

/**
 * `left=right` as the whole file once logged "Scope is not lastParent's parent" while resolving `right`. Nothing
 * binds `right`, so it resolves to nothing, and doing so must not report.
 */
class Issue692Test : PlatformTestCase() {
    fun testUnresolvedAtTopOfFile() {
        myFixture.configureByFiles("unresolved_at_top_of_file.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)
        assertInstanceOf(elementAtCaret, LeafPsiElement::class.java)

        val parent = elementAtCaret!!.parent
        assertNotNull(parent)
        assertInstanceOf(parent, ElixirIdentifier::class.java)

        val grandParent = parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)
        val grandParentCall = grandParent as Call

        val reference = grandParentCall.reference
        assertNotNull(reference)

        val (resolved, errors) = captureLoggedErrors { reference!!.resolve() }

        assertEmpty(errors)
        assertNull("a read with nothing bound above it resolved to something", resolved)
    }

    override fun getTestDataPath(): String = "testData/org/elixir_lang/reference/callable/issue_692"
}
